rholang/src/main/scala/rholang/rosette/Equivalence.scala

```diff
@@ -0,0 +1,330 @@
+      case _ => false
+    }
+
+  def structurallyEquivalent(env1: DeBruijn, p1: Proc, env2: DeBruijn, p2: Proc): Boolean =
+    (p1,p2) match {
+      case (_ : PNil, _) => structurallyNil(p2)
+      case (_, _ : PNil) => structurallyNil(p1)
+      case (v1: PValue, v2: PValue) => valueEquivalent(env1,v1.value_,env2,v2.value_)
+      case (drop1: PDrop, drop2: PDrop) =>
+        nameEquivalent(env1, drop1.chan_, env2, drop2.chan_)
+      case (lift1: PLift, lift2: PLift) => {
+        nameEquivalent(env1, lift1.chan_, env2, lift2.chan_) &&
+          allStructurallyEquivalent(env1, lift1.listproc_, env2, lift2.listproc_)
+      }
+      case (input1: PInput, input2: PInput) =>
+        allBindsEquivalent (
+          env1, input1.listbind_.asScala.toList,
+          env2, input2.listbind_.asScala.toList
+        ) match {
+          case None => false
+          case Some((names1,names2)) =>
+            allCPatternEquivalent(env1,names1,env2,names2) match {
+              case None => false
+              case Some((newEnv1,newEnv2)) =>
+                structurallyEquivalent(newEnv1, input1.proc_, newEnv2, input2.proc_)
+            }
+        }
+      case (choice1: PChoice, choice2: PChoice) =>
+        choice1.listcbranch_.size() == choice2.listcbranch_.size() && (
+          choice1.listcbranch_.asScala.toList,
+          choice2.listcbranch_.asScala.toList
+          ).zipped.forall {
+            case (ch1: Choice,ch2: Choice) =>
+              allBindsEquivalent (
+                env1, ch1.listbind_.asScala.toList,
+                env2, ch2.listbind_.asScala.toList
+              ) match {
+                case None => false
+                case Some((names1,names2)) =>
+                  allCPatternEquivalent(env1,names1,env2,names2) match {
+                    case None => false
+                    case Some((newEnv1,newEnv2)) =>
+                      structurallyEquivalent(
+                        newEnv1, ch1.proc_,
+                        newEnv2, ch2.proc_
+                      )
+                  }
+                }
+            case _ => false
+          }
+        
+      case (match1: PMatch, match2: PMatch) =>
+        val p1 = match1.proc_
+        val p2 = match2.proc_
+        if (structurallyEquivalent(env1,p1,env2,p2)) {
+          val cases1 = match1.listpmbranch_.asScala.toList
+          val cases2 = match2.listpmbranch_.asScala.toList
+          cases1.size == cases2.size &&
+            (cases1,cases2).zipped.forall {
+              case (pm1: PatternMatch,pm2: PatternMatch) =>
+                val pmp1 = pm1.proc_
+                val pmp2 = pm2.proc_
+                if (structurallyEquivalent(env1,pmp1,env2,pmp2)) {
+                  val pat1 = pm1.ppattern_
+                  val pat2 = pm2.ppattern_
+                  (pat1,pat2) match {
+                    case (val1: PPtVal, val2: PPtVal) =>
+                      val1.valpattern_ == val2.valpattern_
```

Comment:
When you go to port this to the mercury grammar, you'll not that there are no patterns. This is on purpose, because a pattern *is* a process with (0 or more) free variables/wildcards.